src/Compilers/CSharp/Test/Emit2/Semantics/LockTests.cs

```diff
@@ -0,0 +1,1797 @@
+                    public Scope EnterLockScope()
+                    {
+                        Console.Write("E");
+                        return new Scope();
+                    }
+
+                    public ref struct Scope
+                    {
+                        public void Dispose()
+                        {
+                            Console.Write("D");
+                        }
+                    }
+                }
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "ELD", verify: Verification.FailsILVerify);
+        verifier.VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void InternalScope()
+    {
+        var source = """
+            System.Threading.Lock l = new();
+            lock (l) { }
+
+            namespace System.Threading
+            {
+                public class Lock
+                {
+                    public Scope EnterLockScope() => new Scope();
+
+                    internal ref struct Scope
+                    {
+                        public void Dispose() { }
+                    }
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (2,7): error CS0656: Missing compiler required member 'System.Threading.Lock.EnterLockScope'
+            // lock (l) { }
+            Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "l").WithArguments("System.Threading.Lock", "EnterLockScope").WithLocation(2, 7),
+            // (8,22): error CS0050: Inconsistent accessibility: return type 'Lock.Scope' is less accessible than method 'Lock.EnterLockScope()'
+            //         public Scope EnterLockScope() => new Scope();
+            Diagnostic(ErrorCode.ERR_BadVisReturnType, "EnterLockScope").WithArguments("System.Threading.Lock.EnterLockScope()", "System.Threading.Lock.Scope").WithLocation(8, 22));
+    }
+
+    [Fact]
+    public void GenericLock()
+    {
+        var source = """
+            static class Program
+            {
+                static void Main()
+                {
+                    System.Threading.Lock<int> l = new();
+                    lock (l) { System.Console.Write("L"); }
+                }
+            }
+
+            namespace System.Threading
+            {
+                public class Lock<T>
+                {
+                    public Scope EnterLockScope()
+                    {
+                        Console.Write("E");
+                        return new Scope();
+                    }
+
+                    public ref struct Scope
+                    {
+                        public void Dispose()
+                        {
+                            Console.Write("D");
+                        }
+                    }
+                }
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "ELD", verify: Verification.FailsILVerify);
+        verifier.VerifyDiagnostics();
```

Comment:
>verifier.VerifyDiagnostics(); [](http://example.com/codeflow?start=8&length=29)

Success scenario should be executed. The generic scenario is interesting enough to actually check the IL, I think.